melange: replace `module_system` and `javascript_extension`

anmonteiro opened the pull request:
- replaces `(module_system ...)` and `(javascript_extension ..)` with a single field:

```clj
(module_systems
 commonjs ;; single module system, default to .js
 (commonjs bs.js)
 (es6 mjs))
```

- allows emitting different module extensions to the same directory and have a single `melange.emit` stanza


### TODO

- [x] fix previous tests

anmonteiro committed:
melange: replace `module_system` and `javascript_extension`

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro committed:
fix: previous tests

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

jchavarri commented on test/blackbox-tests/test-cases/melange/private.t/inside/dune:
```diff
@@ -3,5 +3,5 @@
  (alias melange)
- (module_system es6)
- (javascript_extension mjs))
+ (module_systems
```
I wonder if it'd make sense to add `module_system` in singular as well (in the same vein as `executables` and `executable` stanzas). If it's not very expensive to implement and maintain of course.

anmonteiro replied:
I'm not so sure we should add 2 ways of doing the same thing.

jchavarri replied:
@rgrinberg It is the same with `executables` and `executable` right? I just find weird to write `module_systems` and then only pass one.

anmonteiro replied:
The benefit of `executables` is that you can reuse the module group and the set of library dependencies. What would we gain from having 2 ways of doing it here?

rgrinberg replied:
I think it's more similar to `foreign_stubs` or the `modules` field. I'd leave it as is.

jchavarri commented on src/dune_rules/melange/melange.ml:
```diff
@@ -5,6 +5,8 @@ module Module_system = struct
     | Es6
     | CommonJs
 
+  let default = (CommonJs, ".js")
```
Should default be es6? cc @davesnx

anmonteiro replied:
I'd rather be conservative here and default to CommonJS, which also aligns with the melange defaults.

davesnx replied:
For now, I agree we should default to cjs, since ES6 module is getting stable across the ecosystem.

davesnx commented:
If I get this correct, you can have any number of extensions?

```
# generate 4 files .js, bs.js, cjs.js, .mjs
(module_systems
 commonjs
 (commonjs bs.js)
 (commonjs cjs.js)
 (es6 mjs))
``` 

While the default is `(module_systems commonjs)`.

anmonteiro commented:
@davesnx yes, that's correct.

rgrinberg commented on src/dune_rules/melange/melange.mli:
```diff
@@ -5,6 +5,8 @@ module Module_system : sig
     | Es6
     | CommonJs
 
+  val default : t * string
```
Could you introduce a type alias` Filename.Extension.t` and use it here along with some other places in this PR? Otherwise, the code is a bit hard to read.

jchavarri commented on src/dune_rules/melange/melange_stanzas.ml:
```diff
@@ -81,16 +102,15 @@ module Emit = struct
-           (enum [ ("es6", Melange.Module_system.Es6); ("commonjs", CommonJs) ])
+       and+ module_systems =
+         field "module_systems" module_systems
+           ~default:[ Melange.Module_system.default ]
```
If we add default now, will we get trapped down the line if we ever want to change it?

anmonteiro replied:
I think that can probably be done safely if we bump the melange extension to e.g. `0.2`

jchavarri commented on test/blackbox-tests/test-cases/melange/copy-files-include-subdirs.t:
```diff
@@ -51,3 +51,4 @@ Now add include_subdirs unqualified to show issue
   $ node $src
   hello from file
   
+
```
Why the extra lines?

anmonteiro replied:
it's probably output from the file we print to stdout.

rgrinberg commented on src/dune_rules/melange/melange_stanzas.ml:
```diff
@@ -51,14 +50,36 @@ module Emit = struct
+              (module_system, (loc, ext)))
+      in
+      let (_ : String.Set.t), module_systems =
+        List.fold_map module_systems ~init:String.Set.empty
```
If you were to use `String.Map.of_list_map`, you would be able to improve the error message by including both locations.

jchavarri commented on test/blackbox-tests/test-cases/melange/javascript-extension.t:
```diff
@@ -16,8 +16,7 @@ Can use extension with dots
   >  (alias melange)
-  >  (module_system commonjs)
-  >  (javascript_extension bs.js))
+  >  (module_systems (commonjs bs.js)))
```
I can't comment on the actual line 1, but it includes a description that is now obsolete:

https://github.com/ocaml/dune/blob/955fab5f9bced5987da8265b16f4ca2bcafcfd03/test/blackbox-tests/test-cases/melange/javascript-extension.t#L1

I also wonder if this whole test file should be removed and integrated into the new `modules-system.t`

rgrinberg commented:
We're lacking a test with the same module system being used with different extensions.

jchavarri reviewed (changes requested):
I think the tests should be consolidated before merging, otherwise looks great!

anmonteiro commented:
> We're lacking a test with the same module system being used with different extensions.

The added `module-systems.t` test has it:

```
  >  (module_systems (commonjs bs.js) commonjs (es6 bs.js)))
```

anmonteiro committed:
Address code review

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro commented:
Addressed code review:

- added `Filename.Extension.t`
- switched to using `String.Map.to_list_map` and 2 locations for the repeated extensions
- consolidated test files into one

jchavarri commented on test/blackbox-tests/test-cases/melange/module-systems.t:
```diff
@@ -0,0 +1,136 @@
+  File "dune", line 4, characters 26-29:
+  4 |  (module_systems commonjs es6))
+                                ^^^
+  Error: JavaScript extension .js appears more than once:
```
The wording might be confusing as the extension is not part of the dune config.

anmonteiro replied:
what do you think about [aed2bec](https://github.com/ocaml/dune/pull/7193/commits/aed2becec5fd5331c0b5152ecb65ce9bcd35b49a)?

jchavarri replied:
I think it's way better, thanks!

anmonteiro committed:
test: fix merge

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro committed:
fix(test): provide hints

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

rgrinberg committed:
_

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
_

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>
